traffic_ops/app/db/reencrypt/reencrypt.go

```diff
@@ -0,0 +1,370 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"crypto/aes"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/apache/trafficcontrol/lib/go-util"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/lib/pq"
+)
+
+const PROPERTIES_FILE = "./reencrypt.conf"
+
+func main() {
+	previousKeyLocation := flag.String("previous-key", "/opt/traffic_ops/app/conf/aes.key", "(Optional) The file path for the previous base64 encoded AES key. Default is /opt/traffic_ops/app/conf/aes.key.")
+	newKeyLocation := flag.String("new-key", "/opt/traffic_ops/app/conf/new.key", "(Optional) The file path for the new base64 encoded AES key. Default is /opt/traffic_ops/app/conf/new.key.")
+	cfg := flag.String("cfg", PROPERTIES_FILE, "(Optional) The path for the configuration file. Default is "+PROPERTIES_FILE+".")
+	help := flag.Bool("help", false, "(Optional) Print usage information and exit.")
+	flag.Parse()
+
+	if *help {
+		flag.Usage()
+		os.Exit(0)
+	}
+
+	newKey, err := readKey(*newKeyLocation)
+	if err != nil {
+		die("reading new-key: " + err.Error())
+	}
+
+	previousKey, err := readKey(*previousKeyLocation)
+	if err != nil {
+		die("reading previous-key: " + err.Error())
+	}
+
+	dbConfBytes, err := ioutil.ReadFile(*cfg)
+	if err != nil {
+		die("reading db conf '" + *cfg + "': " + err.Error())
+	}
+
+	pgCfg := Config{}
+	err = json.Unmarshal(dbConfBytes, &pgCfg)
+	if err != nil {
+		die("unmarshalling '" + *cfg + "': " + err.Error())
+	}
+
+	sslStr := "require"
+	if !pgCfg.SSL {
+		sslStr = "disable"
+	}
+	db, err := sqlx.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&fallback_application_name=trafficvault", pgCfg.User, pgCfg.Password, pgCfg.Hostname, pgCfg.Port, pgCfg.DBName, sslStr))
+	if err != nil {
+		die("opening database: " + err.Error())
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		die(fmt.Sprintf("transaction begin failed %v %v ", err, tx))
+	}
+	defer tx.Commit()
+
+	if err = reEncryptSslKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting SSL Keys: " + err.Error())
+	}
+	if err = reEncryptUrlSigKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting URL Sig Keys: " + err.Error())
+	}
+	if err = reEncryptUriSigningKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting URI Signing Keys: " + err.Error())
+	}
+	if err = reEncryptDNSSECKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting DNSSEC Keys: " + err.Error())
+	}
+
+	if err = updateKeyFile(previousKey, *previousKeyLocation, newKey); err != nil {
```

Comment:
So after trying this out, I think this behavior might be a little surprising to the user. If there was only one copy of the original key in existence (which would be crazy I know), running this tool would erase it, making it impossible to roll back. Should we remove this step? It's easy enough to manually `mv` the files after.

Reply:
It saves it out in the same location as the original key but with the name `savedEncryptionKey-` and then the date and time of re-encryption. so if you do only have one copy then it will still save it

Reply:
on second thought i also dont really care if it doesnt save it out / replace the old key. i thought it would be nice so you can stop TO, re-encrypt, and restart TO without much thought and save the annoyance of restarting TO with the wrong key location but as you said its not much effort to just `mv` it later. so i dont have a strong opinion either way if we want to change it back

Reply:
Oh, I remember reading that in the code on my first review but forgot about that. I think doing an `mv` after is ok, then we don't also have to remove old backup key files later. If we were to automate this, we'd probably wrap it in a script that generates a new key and handles key backup and other stuff. Then re-encryption would basically be a single button.